Do not remove calls to functions without a body during inlining

danpoe opened the pull request:
When using ``--inline`` or ``--function-inline`` in goto-instrument, calls to functions without a body are now kept (partial inlining already keeps them).

Additionally there's now a separate analysis pass ``--remove-calls-nobody`` which removes calls to functions without a body and replaces them by a nondet assignment to the return variable and evaluations of the call arguments.

tautschnig commented:
Throughout this PR: shouldn't it be `no_body` (and `no-body`) instead of `nobody`?

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -683,6 +686,10 @@ int goto_instrument_parse_optionst::doit()
 
+      status() << "Removing calls to functions without a body" << eom;
+      remove_calls_nobodyt rcn;
+      rcn.remove_calls_nobody(goto_functions);
```
Probably more "cprovery" to use `operator()` so this would become:

```C++
remove_calls_nobodyt remove_calls_nobody;
remove_calls_nobody(goto_functions);
```

thk123 commented on regression/goto-instrument/remove-calls-nobody2/main.c:
```diff
@@ -0,0 +1,20 @@
+
```
Nit: blank line

thk123 commented on regression/goto-instrument/remove-calls-nobody1/main.c:
```diff
@@ -0,0 +1,18 @@
+
```
Nit: blank line

thk123 commented on regression/goto-instrument/inline_17/main.c:
```diff
@@ -0,0 +1,11 @@
+
```
Nit: blank line

thk123 commented on regression/goto-instrument/inline_16/main.c:
```diff
@@ -0,0 +1,11 @@
+
```
Nit: blank line

thk123 commented on regression/goto-instrument/inline_17/main.c:
```diff
@@ -0,0 +1,11 @@
+{
+  func1();
+}
+
```
Nit: blank line

thk123 commented on regression/goto-instrument/inline_16/main.c:
```diff
@@ -0,0 +1,11 @@
+  func1();
+  ret = func2();
+}
+
```
Nit: blank line

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1075,27 +1082,21 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   }
 
-  // now do full inlining, if requested
-  if(cmdline.isset("inline"))
```
Am I reading this right? There were two identical blocks for dealing with `cmdline.isset("inline")`?

danpoe replied:
Yes

thk123 commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -76,7 +76,8 @@ Author: Daniel Kroening, kroening@kroening.com
   "(undefined-function-is-assume-false)" \
-  "(remove-function-body):"
+  "(remove-function-body):" \
+  "(remove-calls-nobody)"
```
This flag is missing documentation, also if possible could you move the defines for it into `remove_calls_nobody.h` as described here: https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD#L114

thk123 commented on src/goto-programs/goto_inline_class.cpp:
```diff
@@ -450,7 +450,7 @@ void goto_inlinet::insert_function_body(
 }
 
-void goto_inlinet::insert_function_nobody(
+void goto_inlinet::handle_function_nobody(
```
Now would be a good time to add the doxygen header to this function as it isn't clear what handling a function without a body entails.

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+
+#include "remove_calls_nobody.h"
+
+void remove_calls_nobodyt::remove_call_nobody(
```
Could you add Doxygen comments for all the functions?

thk123 replied:
Instead of all methods being called `remove_call_nobody`, perhaps this could be called `generate_nondet_method_stub`

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+
+  // evaluate function arguments -- they might have
+  // pointer dereferencing or the like
+  forall_expr(it, arguments)
```
I think this can be done using a range based for which is now preferred:

```C++
for(const exprt &argument : arguments)
```

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+  forall_expr(it, arguments)
+  {
+    goto_programt::targett t=tmp.add_instruction();
+    t->make_other(code_expressiont(*it));
```
This is probably me not understanding something - but what does making an other expression do? Don't we need some kind of nondet modification if it is a pointer?

danpoe replied:
An other expression just amounts to evaluating the expression. For example, for a call ``func(a[i])`` we insert an evaluation of ``a[i]`` so we can still detect out of bounds errors on the array access. The intention here is not to soundly model everything the function might do with say a pointer argument but just to "simulate" the program evaluating the arguments to the call before making the call.

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+
+    if(!may_have_body)
+    {
+      remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
```
Is it OK to modify `goto_program` while we have an interator pointing into it?

NlightNFotis replied:
This is also something that I want to know.

danpoe replied:
Yes, that's fine. Since it's a ``std::list`` of instructions all the iterators to elements in the list remain valid when changing the list.

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+  dest.destructive_insert(target, tmp);
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(
```
Instead of all being called `remove_calls_nobody` maybe this could be called `remove_no_body_calls_from_function`

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+
+void remove_calls_nobodyt::remove_calls_nobody(goto_functionst &goto_functions)
+{
+  Forall_goto_functions(f_it, goto_functions)
```
Range based for:

```C++
for(auto &function_entry : goto_functions)
```

danpoe replied:
This would need to be ``for(auto &function_entry : goto_functions.function_map)`` which I'm not sure would be preferable.

thk123 replied:
I would prefer it since I think it is more idiomatic, someone new to the code doesn't need to check what Forall_goto_functions does, but I don't feel strongly about it - as you prefer.

tautschnig replied:
While I do agree, I am wondering whether there should be a concerted attempt to replace all occurrences of `Forall_goto_functions` in a consistent way instead of risking to introduce a multitude of variations expressing the same? Thoughts?

thk123 replied:
I agree in theory; in practise I think big sweeping changes tend to takes ages/never get merged/break peoples PRs etc. Small, incremental improvements in code people are actually modifying nicely prioritises code people are likely to touch and actually gets done.

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+  for(goto_programt::targett it=goto_program.instructions.begin();
+      it!=goto_program.instructions.end();) // no it++
+  {
+    if(!it->is_function_call())
```
I'd be tempted to break the logic of this loop up into two parts:

```C++
bool is_opaque_function_call(const goto_programt::target instruction, const goto_functions &goto_functions)
{
  if(!instruction->is_function_call()) 
    return false;
...
if(!goto_function.body_available())
  return true;
}
```

And then this function becomes:

```C++
for(goto_programt::targett it=goto_program.instructions.begin();
    it!=goto_program.instructions.end();) // no it++
{
  if(is_opaque_function_call(it, goto_function)
  {
    remove_call_nobody(goto_program, it, cfc.lhs(), cfc.arguments());
  }
  else
  {
    it++;
  }
}
```

This way it is easy to see what the loop is doing and when the iterator gets advanced

danpoe replied:
Good suggestion. It made the code a lot clearer.

thk123 commented on src/goto-programs/remove_calls_nobody.cpp:
```diff
@@ -0,0 +1,116 @@
+  }
+}
+
+void remove_calls_nobodyt::remove_calls_nobody(goto_functionst &goto_functions)
```
As suggested earlier, this would become `operator()(goto_functionst &goto_functions)`

thk123 commented on src/goto-programs/remove_calls_nobody.h:
```diff
@@ -0,0 +1,35 @@
+    goto_programt &goto_program,
+    const goto_functionst &goto_functions);
+
+  void remove_calls_nobody(goto_functionst &goto_functions);
```
I think this method is the only one that should be public as the others are utilities that only make sense to be called in the specific way that this one calls them.

thk123 commented on regression/goto-instrument/inline_17/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+func2\(\)
+--
+^warning: ignoring
```
Should this test check `func1` isn't called?

NlightNFotis replied:
I agree with this. `func2\(\)` appears to me like it would be in the file even if it `func1` isn't actually inlined.

danpoe replied:
I think it's fine. ``--function-inline func1`` inlines all the calls that ``func1`` makes but the call to ``func1`` itself remains. I've also added a test now that ``func1`` is called.

thk123 commented on regression/goto-instrument/inline_09/test.desc:
```diff
@@ -1,6 +1,6 @@
 CORE
 main.c
---inline
+--inline --remove-calls-nobody
```
Could we add to this test a requirement that there is no call to `f`

danpoe replied:
Good point.

danpoe commented:
I think I've addressed all the comments. I didn't add new tests tough as I think the current ones are sufficient, since we're not aiming to add nondet assignments to things a pointer might point to but simply to evaluate the arguments.

peterschrammel commented on src/goto-programs/goto_inline.cpp:
```diff
@@ -48,40 +48,40 @@ void goto_inline(
 
   typedef goto_functionst::goto_functiont goto_functiont;
 
-    // find entry point
```
The style clean-ups should probably go into a separate commit.

danpoe commented:
Ok, I've squashed the last commit. I've also removed the function ``handle_function_no_body()`` as it no longer warrants a separate function.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -696,6 +700,10 @@ int goto_instrument_parse_optionst::doit()
+      status() << "Removing calls to functions without a body" << eom;
+      remove_calls_no_bodyt remove_calls_no_body;
+      remove_calls_no_body(goto_functions);
+
```
Why is this specifically done here - is this the only place calling `goto_inline`?

danpoe replied:
Yes, this is the only analysis that specifically calls ``goto_inline`` before.

tautschnig commented on src/goto-programs/remove_calls_no_body.cpp:
```diff
@@ -0,0 +1,137 @@
+  {
+    goto_functiont &goto_function=f_it->second;
+
+    if(goto_function.body_available())
```
I don't think you even need this check - an empty body will simply make the loop terminate straight away.

danpoe replied:
Right.

tautschnig replied:
Was there an intention to make a change?

tautschnig commented on src/goto-programs/remove_calls_no_body.h:
```diff
@@ -0,0 +1,46 @@
+class remove_calls_no_bodyt
+{
+protected:
+  typedef goto_functionst::goto_functiont goto_functiont;
```
I am wondering whether this typedef is of much value - it seems to be used twice, and one of those you could even eliminate given my earlier comment.

danpoe replied:
Ok, I'm removing it.

tautschnig replied:
Happy to approve once the changes have been pushed.

danpoe replied:
Ok, I've pushed the changes.

tautschnig replied:
Is that line a left-over to be removed?

kroening commented:
Needs rebase, but can then go!

danpoe commented:
@martin-cs As far as I remember we produced the Toyota releases by merging together a bunch of branches, this one being one of them. Or has the process changed meanwhile? Which branch and which patches are you referring to?

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -990,7 +992,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   if(cmdline.isset("inline"))
   {
-    do_indirect_call_and_rtti_removal();
+    do_indirect_call_and_rtti_removal(/*force=*/true);
```
It might be better change the API to use an enum to make clear what the value means?

danpoe replied:
I think having boolean flags for options should be fine.

tautschnig replied:
It surely is, but the fact that you felt compelled to add a comment suggests that an enum (with just two values!) could be helpful for people reading the code.

danpoe replied:
I agree that it would be more readable. I'm slightly concerned about consistency with the rest of the code base though as having two-value enums for boolean flags isn't done commonly there. I believe I've added the comment explaining the argument here just randomly (I think it's also the case for lots of other function calls with boolean literals as arguments that it's not clear what the flag does unless one inspects the declaration/definition). I'll remove the comment from the call.

tautschnig replied:
I know that #1547 introduced one of those, and there was some PR (not sure whether that one or another) that got me very confused initially. Indeed it's not yet common practice, but likely we should look into adopting it. Because I'm in the mood: #1854.

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1109,29 +1111,15 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+    goto_functions.compute_loop_numbers();
   }
 
+
```
Nit pick: blank line seems unnecessary?

tautschnig commented on src/goto-programs/goto_inline_class.h:
```diff
@@ -6,7 +6,6 @@ Author: Daniel Kroening, kroening@kroening.com
 
 \*******************************************************************/
 
-
```
Nit pick: change seems not too important?

martin-cs commented:
On Thu, 2018-02-15 at 09:05 -0800, danpoe wrote:
> @martin-cs As far as I remember we produced the Toyota releases by merging together a bunch of branches, this one being one of them. Or has the process changed meanwhile? Which branch and which patches are you referring to?
@danpoe A few things have changed.  Let's get this merged and then have a look.
